### PR TITLE
NFC: Add comments to _IndexedComponent_slice_iter and Reference

### PR DESCRIPTION
**Pull request by Robbybp**
## Summary/Motivation:
The code for these objects is dense and sometimes hard to follow. These comments hopefully make it less so (at least for somebody already with a partial understanding of these classes). They reflect only my current level of understanding, and include questions where I am confused.

## Changes proposed in this PR:
- Comments in `_IndexedComponent_slice_iter`, `_Reference*`, and `Reference`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/base/indexed_component_slice.py [hunk 1]
```diff
@@ -399,13 +421,20 @@ def __init__(self, component_slice, advance_iter=_advance_iter,
...
                 *call_stack[0][1], iter_over_index=self._iter_over_index)
+            # call_stack[0][1] is a (fixed, sliced, ellipsis) tuple, where
+            # fixed and sliced are dicts.
         elif call_stack[0][0] == IndexedComponent_slice.set_item:
             assert call_stack_len == 1
             # defer creating the iterator until later
             self._iter_stack[0] = _NotIterable # Something not None
+            # Initializing to _NotIterable is necessary to avoid an
+            # error when we walk to the base of the call stack, but why
+            # would `set_item` appear at the base of the call stack?
```
Comment (Robbybp) [inline pyomo/core/base/indexed_component_slice.py:437]: Still don't really understand why iterator creation is deferred. Why would `set_item` appear on the base of the stack.

**Review by blnicho — COMMENTED** [review]
I found a few typos but otherwise I think this looks great. I don't have any answers for the questions you added, @jsiirola will have to look at them.

File: pyomo/core/base/indexed_component_slice.py [hunk 2]
```diff
@@ -491,10 +579,16 @@ def __next__(self):
...
+                        # Unclear to me when this would be necessary...
+                        # -RBP
                         self._iter_stack[idx] = None
                 elif _call[0] == IndexedComponent_slice.call:
                     try:
+                        # Assume the callable "comp" in our hierarchy
+                        # returns a component:
                         _comp = _comp( *(_call[1]), **(_call[2]) )
+                        # The only callable "comp" here should be
+                        # the `component` method, right? -RBP
```
Comment (jsiirola) [inline pyomo/core/base/indexed_component_slice.py:591]: No; consider:
```python
m.x[:].fix(5)
```

File: pyomo/core/base/indexed_component_slice.py [hunk 3]
```diff
@@ -654,8 +759,16 @@ def get_last_index(self):
             return ans
 
     def get_last_index_wildcards(self):
+        """
+        Get a tuple of the values in the wildcard positions for 
+        the most recent indices checked by each _slice_generator
```
Comment (jsiirola) [inline pyomo/core/base/indexed_component_slice.py:764]: "checked" is probably the wrong verb.  Wouldn't "returned" be more accurate?

File: pyomo/core/base/indexed_component_slice.py [hunk 4]
```diff
@@ -654,8 +759,16 @@ def get_last_index(self):
...
+        in the iter stack.
+        """
+        # This method is how we iterate over keys.
         ans = sum(
             ( tuple( x.last_index[i]
+                # last_index is the most recent index checked
```
Comment (jsiirola) [inline pyomo/core/base/indexed_component_slice.py:770]: Again, "returned" and not "checked"?

File: pyomo/core/base/reference.py [hunk 5]
```diff
@@ -100,22 +107,38 @@ def __call__(self, _slice):
...
         try:
+            # Here we assemble the index we will actually use to access
+            # the component.
             idx = tuple(
                 _slice.fixed[i] if i in _slice.fixed else self.key.pop(0)
                 for i in range(idx_count))
+            # _slice corresponds to some sliced entry in the call/iter stacks
+            # that contains the information describing the slice.
+            # Here we fill in an index with the fixed indices from the slice
+            # the wildcard indices from the provided key.
```
Comment (jsiirola) [inline pyomo/core/base/reference.py:118]: missing: "and" ("fixed indices from the slice *and* the wildcard indices from")

File: pyomo/core/base/reference.py [hunk 6]
```diff
@@ -348,62 +387,103 @@ def _get_iter(self, _slice, key):
...
+            # of the first coordinate of a set.
+            # This function seems to only make sense if 
+            # normalize_index.flatten is True.
             wildcard_sets = {}
+            # `wildcard_sets` maps position in the "subsets list"
+            # to its set if that set is a wildcard?
             for j,s in enumerate(level.component.index_set().subsets()):
+                # Iterate over the sets that could possibly be wildcards
                 if s is UnindexedComponent_set:
+                    # Why is this treated as a wildcard? (How is this used?)
```
Comment (jsiirola) [inline pyomo/core/base/reference.py:415]: You can hit this weirdness when scalars are treated as indexed components (explicitly or implicitly):
```python
m = ConcreteModel()
m.x = Var()
m.r1 = Reference(m.x)
m.r2 = Reference(m.x[...])

m.b = Block()
m.b.y = Var()
m.r3 = Reference(m.b[:].m.y[:])
```

File: pyomo/core/base/reference.py [hunk 7]
```diff
@@ -348,62 +387,103 @@ def _get_iter(self, _slice, key):
...
     for i, level in enumerate(iter_stack):
         if level is not None:
             offset = 0
+            # `offset` is the position, within the total index tuple,
+            # of the first coordinate of a set.
+            # This function seems to only make sense if 
+            # normalize_index.flatten is True.
             wildcard_sets = {}
+            # `wildcard_sets` maps position in the "subsets list"
+            # to its set if that set is a wildcard?
```
Comment (jsiirola) [inline pyomo/core/base/reference.py:411]: Yes:
```
# `wildcard_sets` maps a position in the current level's 
# "subsets list" to its corresponding set if that set is a wildcard.
```

File: pyomo/core/base/reference.py [hunk 8]
```diff
@@ -348,62 +387,103 @@ def _get_iter(self, _slice, key):
...
+                    # Why is this treated as a wildcard? (How is this used?)
                     wildcard_sets[j] = s
                     offset += 1
                     continue
                 if s.dimen is None:
                     return None
+                # `wild` is the number of coordinates of this set (which
+                # may be multi-dimensional) that have been sliced.
                 wild = sum( 1 for k in range(s.dimen)
                             if k+offset not in level.fixed )
```
Comment (jsiirola) [inline pyomo/core/base/reference.py:424]: We should go ahead and rename this variable `wildcard_count`

File: pyomo/core/base/reference.py [hunk 9]
```diff
@@ -348,62 +387,103 @@ def _get_iter(self, _slice, key):
...
+    # dict maps position within that level's component's "subsets"
+    # list to the set at that position if it is a wildcard set.
     tmp = [None]*len(iter_stack)
     for i, level in enumerate(iter_stack):
         if level is not None:
             offset = 0
+            # `offset` is the position, within the total index tuple,
+            # of the first coordinate of a set.
+            # This function seems to only make sense if 
+            # normalize_index.flatten is True.
```
Comment (jsiirola) [inline pyomo/core/base/reference.py:408]: This comment should probably be promoted to to the method documentation and not buried down here.

File: pyomo/core/base/reference.py [hunk 10]
```diff
@@ -348,62 +387,103 @@ def _get_iter(self, _slice, key):
...
+                # `wild` is the number of coordinates of this set (which
+                # may be multi-dimensional) that have been sliced.
                 wild = sum( 1 for k in range(s.dimen)
                             if k+offset not in level.fixed )
+                # `k+offset` is a position in the "total" (flattened)
+                # index tuple.
+                # All the _slice_generator's information is in terms
+                # of this total index tuple.
                 if wild == s.dimen:
+                    # Every coordinate of set is covered by a wildcard.
```
Comment (jsiirola) [inline pyomo/core/base/reference.py:430]: "Every coordinate of *this subset* is covered"

File: pyomo/core/base/reference.py [hunk 11]
```diff
@@ -348,62 +387,103 @@ def _get_iter(self, _slice, key):
...
-    # if we have already decided that there isn't a comon index for the
+    # if we have already decided that there isn't a common index for the
     # slices, there is nothing more we can do.  Bail.
     if index is None:
         return index
 
     # Walk the iter_stack that led to the current item and try to
     # identify the component wildcard sets
+    #
+    # `tmp` will be a list of dicts mirroring the iter_stack. Each
```
Comment (jsiirola) [inline pyomo/core/base/reference.py:398]: We should probably rename `tmp` to be something more meaningful - as it holds the ultimate answer!

**Review by jsiirola — CHANGES_REQUESTED** [review]
Overall, this is great documentation.  I (think) I answered all your embedded questions.  There is one thing that we really need to resolve before merging (directionality of stacks).

File: pyomo/core/base/indexed_component_slice.py [hunk 12]
```diff
@@ -420,33 +449,80 @@ def next(self):
...
+            # function encountered a slice. Previous iterators could
+            # be active as well, but we wait to deal with them until
+            # we have (temporarily) exhausted the "highest level"
+            # iterator.
             while self._iter_stack[idx] is None:
                 idx -= 1
             # Get the next element in the deepest active slice
             try:
                 if self._iter_stack[idx] is _NotIterable:
+                    # When does this happen?
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply pyomo/core/base/indexed_component_slice.py:471]: Am I correct to assert that `_ReferenceDict.__setitem__` only gets hit if we are directly working with a `_ReferenceDict` object directly, rather than with the reference itself? E.g. 
```python
ref = Reference(m.x[:])
ref[1] = 10
```
would not hit `_ReferenceDict.__setitem__` and therefore would not hit the `_NotIterable` code we are talking about.

File: pyomo/core/base/indexed_component_slice.py (hunk 3 above)
Comment (Robbybp) [reply pyomo/core/base/indexed_component_slice.py:764]: The `_slice_generator` returns components, not indices (or at least advancing it does), so I'm averse to "returned." I'll go with "used" for now.

File: pyomo/core/base/indexed_component_slice.py (hunk 4 above)
Comment (Robbybp) [reply pyomo/core/base/indexed_component_slice.py:770]: I prefer "used" to either.

File: pyomo/core/base/reference.py (hunk 11 above)
Comment (Robbybp) [reply pyomo/core/base/reference.py]: Have renamed to `wildcard_stack`.

File: pyomo/core/base/indexed_component_slice.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/core/base/indexed_component_slice.py]: Yes.  I believe that is correct.

File: pyomo/core/base/indexed_component_slice.py (hunk 3 above)
Comment (jsiirola) [reply pyomo/core/base/indexed_component_slice.py]: Good point.  What about:
```
Return a tuple of the values in the wildcard positions for 
the index corresponding to the component most recently
returned by each _slice_generator.
```

**Review by jsiirola — APPROVED** [review]

**Comment by Robbybp** [comment]
I somehow introduced 700 files of changes by merging the wrong branch here. Not sure how to fix this. I've cherry-picked the changes I want and will re-open in a new PR.

**Comment by jsiirola** [comment]
I think you can close this PR, checkout fa6876aee35d6069f545d1515f8f96eb7c8a89b2 (in HEADless mode), create a new branch, push, and open a new PR.

**Comment by Robbybp** [comment]
Already made the necessary changes by cherry-picking. Closing this PR.
